File: apeiria/ai/webui/routes/sources.py
"""AI source admin routes."""

from __future__ import annotations

from typing import TYPE_CHECKING, Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status

from apeiria.ai.admin.service import (
    AISourceDeleteBlockedError,
    ai_admin_service,
)
from apeiria.ai.webui.schemas import (
    AISourceItem,
    AISourcePresetItem,
    AISourceUpsertRequest,
)
from apeiria.ai.webui.support import (
    to_ai_source_item,
    to_ai_source_preset_item,
)
from apeiria.webui.auth import require_control_panel

if TYPE_CHECKING:
    from apeiria.access.principal import AuthSession


router = APIRouter()


def _actor_username_from_claims(session: "AuthSession") -> str | None:
    username = session.username.strip()
    return username or None


@router.get("/source-presets", response_model=list[AISourcePresetItem])
async def list_ai_source_presets(
    _: Annotated[Any, Depends(require_control_panel)],
) -> list[AISourcePresetItem]:
    return [
        to_ai_source_preset_item(item)
        for item in ai_admin_service.list_source_presets()
    ]


@router.get("/sources", response_model=list[AISourceItem])
async def list_ai_sources(
    _: Annotated[Any, Depends(require_control_panel)],
) -> list[AISourceItem]:
    items = await ai_admin_service.list_sources()
    return [to_ai_source_item(item) for item in items]


@router.post("/sources", response_model=AISourceItem)
async def create_ai_source(
    payload: AISourceUpsertRequest,
    session: Annotated["AuthSession", Depends(require_control_panel)],
) -> AISourceItem:
    item = await ai_admin_service.create_source(
        name=payload.name,
        capability_type=payload.capability_type,
        preset_type=payload.preset_type,
        api_base=payload.api_base,
        api_key_env_name=payload.api_key_env_name,
        enabled=payload.enabled,
        timeout_seconds=payload.timeout_seconds,
        custom_headers=payload.custom_headers,
        extra_config=payload.extra_config,
        actor_username=_actor_username_from_claims(session),
    )
    return to_ai_source_item(item)


@router.put("/sources", response_model=AISourceItem | None)
async def update_ai_source(
    payload: AISourceUpsertRequest,
    session: Annotated["AuthSession", Depends(require_control_panel)],
) -> AISourceItem | None:
    if not payload.source_id:
        return None
    item = await ai_admin_service.update_source(
        source_id=payload.source_id,
        name=payload.name,
        capability_type=payload.capability_type,
        preset_type=payload.preset_type,
        api_base=payload.api_base,
        api_key_env_name=payload.api_key_env_name,
        enabled=payload.enabled,
        timeout_seconds=payload.timeout_seconds,
        custom_headers=payload.custom_headers,
        extra_config=payload.extra_config,
        actor_username=_actor_username_from_claims(session),
    )
    return to_ai_source_item(item) if item is not None else None


@router.delete("/sources", response_model=bool)
async def delete_ai_source(
    session: Annotated["AuthSession", Depends(require_control_panel)],
    source_id: Annotated[str, Query(min_length=1)],
) -> bool:
    try:
        return await ai_admin_service.delete_source(
            source_id=source_id,
            actor_username=_actor_username_from_claims(session),
        )
    except AISourceDeleteBlockedError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(exc),
        ) from exc


__all__ = ["router"]

File: webui/src/api/ai/sources.ts
import client from '@/api/client'
import type {
  AIBootstrapResponse,
  AIRuntimeStatusResponse,
  AIRuntimeSettingsResponse,
  AIRuntimeSettingsUpdateRequest,
  AISourceApiKeyAction,
  AISourceItem,
  AISourcePresetItem,
} from './types'

export { AIBootstrapResponse, AIRuntimeStatusResponse, AIRuntimeSettingsResponse, AIRuntimeSettingsUpdateRequest }

export function getAIBootstrap() {
  return client.get<AIBootstrapResponse>('/ai/bootstrap')
}

export function getAIRuntimeStatus() {
  return client.get<AIRuntimeStatusResponse>('/ai/runtime-status')
}

export function getAIRuntimeSettings() {
  return client.get<AIRuntimeSettingsResponse>('/ai/runtime-settings')
}

export function updateAIRuntimeSettings(payload: AIRuntimeSettingsUpdateRequest) {
  return client.patch<AIRuntimeSettingsResponse>('/ai/runtime-settings', payload)
}

export function getAISources() {
  return client.get<AISourceItem[]>('/ai/sources')
}

export function getAISourcePresets() {
  return client.get<AISourcePresetItem[]>('/ai/source-presets')
}

export function createAISource(payload: {
  name: string
  capability_type: string
  preset_type: string
  api_base?: string | null
  enabled: boolean
  timeout_seconds?: number | null
  custom_headers?: Record<string, string>
  extra_config?: Record<string, unknown>
  adapter_kind?: string | null
  capability_metadata?: Record<string, unknown>
  default_options?: Record<string, unknown>
  capability_provenance?: Record<string, unknown>
  api_key_action?: AISourceApiKeyAction | null
  api_keys?: string[]
}) {
  return client.post<AISourceItem>('/ai/sources', payload)
}

export function updateAISource(payload: {
  source_id: string
  name: string
  capability_type: string
  preset_type: string
  api_base?: string | null
  enabled: boolean
  timeout_seconds?: number | null
  custom_headers?: Record<string, string>
  extra_config?: Record<string, unknown>
  adapter_kind?: string | null
  capability_metadata?: Record<string, unknown>
  default_options?: Record<string, unknown>
  capability_provenance?: Record<string, unknown>
  api_key_action?: AISourceApiKeyAction | null
  api_keys?: string[]
}) {
  return client.put<AISourceItem | null>('/ai/sources', payload)
}

export function deleteAISource(sourceId: string) {
  return client.delete<boolean>('/ai/sources', {
    params: { source_id: sourceId },
  })
}
